presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java

```diff
@@ -223,13 +226,32 @@ private boolean writeDictionaryRowGroup(Block dictionary, int valueCount, IntBig
         for (int i = 0; valueCount > 0 && i < segments.length; i++) {
             int[] segment = segments[i];
             int positionCount = Math.min(valueCount, segment.length);
-            DictionaryBlock dictionaryBlock = new DictionaryBlock(positionCount, dictionary, segment);
-            directColumnWriter.writeBlock(dictionaryBlock);
-            valueCount -= positionCount;
+            Block block = new DictionaryBlock(positionCount, dictionary, segment);
+
+            while (block != null) {
+                int chunkPositionCount = block.getPositionCount();
+                Block chunk = block.getRegion(0, chunkPositionCount);
 
-            if (directColumnWriter.getBufferedBytes() > maxDirectBytes) {
-                return false;
+                // avoid chunk with huge logical size
+                while (chunkPositionCount > 1 && chunk.getLogicalSizeInBytes() > DIRECT_CONVERSION_CHUNK_MAX_LOGICAL_BYTES) {
+                    chunkPositionCount /= 2;
+                    chunk = chunk.getRegion(0, chunkPositionCount);
+                }
+
+                if (chunkPositionCount < block.getPositionCount()) {
```

Comment:
I'd move this to the bottom of the `while` loop and add a comment like `// restrict block to only unwritten rows`